### PR TITLE
styled-components: Update forwardedAs to vary from as

### PR DESCRIPTION
**Pull request by merrywhether**
This allows the `as` and `forwardedAs` props to vary from each other, which is appropriate because they have different targets.

This also runs prettier (per the doc recommendations) on the `styled-components/` directory in a second commit, generating a lot of visual updates without changing anything. Look at the first commit for the functional changes.

- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [X] Add or edit tests to reflect the change. (Run with `npm test YOUR_PACKAGE_NAME`.)
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [X] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [X] Provide a URL to documentation or source code which provides context for the suggested changes: https://styled-components.com/docs/api#forwardedas-prop
- [X] Include [tests for your changes](https://github.com/DefinitelyTyped/DefinitelyTyped#testing)

### CODE REVIEW

**Comment by danger-public** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## styled-components ([<kbd>unpkg</kbd>](https://unpkg.com/browse/styled-components@latest/))
was missing the following properties:
1. StyleSheetConsumer
1. StyleSheetContext
1. version
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against d04ae31d9268dcc40f6435233d6257424df30d64
</p>

**Comment by merrywhether** [comment]
After wrangling with some of the changes introduced by prettier (it moved some magic comments as it re-broke lines), this should be RFR.

**Comment by merrywhether** [comment]
Any reviewers out there?

**Comment by merrywhether** [comment]
Bump @Igorbek @Igmat @lavoaster @Jessidhia @JKillian @eps1lon @flavordaaave @wagerfield @Lazyuki @mgoszcz2 @danilofuchs

File: types/styled-components/index.d.ts [hunk 1]
```diff
@@ -88,10 +88,11 @@ type WithChildrenIfReactComponentClass<
 
 type StyledComponentPropsWithAs<
     C extends string | React.ComponentType<any>,
+    F extends string | React.ComponentType<any>,
```
Comment (rbuckton) [inline types/styled-components/index.d.ts:109]: Introducing a new type parameter in the middle of existing type parameters is a breaking change, especially for anyone who might have an explicit `StyledCompontsPropsWithAs` reference in their code (a quick search of GitHub for just `StyledComponentPropsWithAs` shows this happens with some regularity).

Would it be feasible to add this as a trailing type parameter with a default `= C` to maintain backwards compatibility? If not, I'd need one of the definition owners to provide feedback before I'd feel comfortable approving this change.

**Review by rbuckton — CHANGES_REQUESTED** [review]

File: types/styled-components/index.d.ts (hunk 1 above)
Comment (merrywhether) [reply types/styled-components/index.d.ts]: Yeah, that seems reasonable. It mangles the signatures a bit, but seems worth it for providing backwards compatibility.

File: types/styled-components/ts3.6/index.d.ts [hunk 2]
```diff
@@ -75,59 +74,34 @@ type StyledComponentPropsWithAs<
     T extends object,
     O extends object,
     A extends keyof any
-> = StyledComponentProps<C, T, O, A> & { as?: C, forwardedAs?: C };
+> = StyledComponentProps<C, T, O, A> & { as?: C; forwardedAs?: C };
```
Comment (rbuckton) [inline types/styled-components/ts3.6/index.d.ts:77]: Is there a reason not to support the same `forwardedAs` variance in the ts3.6 types?

**Review by rbuckton — COMMENTED** [review]
Seems good to me, but I have one other comment before I approve.

File: types/styled-components/ts3.6/index.d.ts (hunk 2 above)
Comment (merrywhether) [reply types/styled-components/ts3.6/index.d.ts:77]: Oh, that's TS@3.6 not Styled Components@3.6. 😞 

Yup, I can update that too.

Comment (merrywhether) [reply types/styled-components/ts3.6/index.d.ts]: Gah, lands squarely in a weird TODO section. Let's see how true that TODO comment is...

**Review by rbuckton — APPROVED** [review]

**Comment by merrywhether** [comment]
Ready to merge

**Pull request merged**